app/questionnaire/router.py

```diff
@@ -258,6 +268,121 @@ def _get_return_to_location_url(
                 "questionnaire.submit_questionnaire", _anchor=return_to_answer_id
             )
 
+    def _get_return_to_for_grand_calculated_summary(
+        self,
+        *,
+        return_to: str | None,
+        return_to_block_id: str | None,
+        location: Location,
+        routing_path: RoutingPath,
+        is_for_previous: bool,
+        return_to_answer_id: str | None = None,
+    ) -> str | None:
+        """
+        Builds the return url for a grand calculated summary,
+        and accounts for it possibly being in a different section to the calculated summaries it references
+        """
+        grand_calculated_summary_section = location.section_id
+        # if return_to_block_id is invalid and no block is found, the section doesn't update but can_access_location will be false, so it still returns None
+        if return_to_block_id and self._schema.get_block(return_to_block_id):
+            # Type ignore: if the block is valid, then we'll be able to find a section for it
+            grand_calculated_summary_section: str = self._schema.get_section_id_for_block_id(return_to_block_id)  # type: ignore
+        if grand_calculated_summary_section != location.section_id:
```

Comment:
If the first if fails, then you should return None. At the moment, it can go into this if then evaluate the path again for no reason. If you guarded this second if, then it would also work as it would return false for can_access_location but I would return soon as you can.

Reply:
Fixed - thanks 👍